Trigger stop action when calling flt() and similar functions

JakubVanek opened the pull request:
Previously only the stop mode was configured in some functions. This however did not trigger the stop action itself.
This commit adds a stop action execution to flt(), coast(), brake(), hold() and stop().
Also, the immediateReturn argument is supported. There is also a change against the leJOS default - the non-parameterized stop() and flt() calls are non-blocking now. This is documented in the leJOS ADR.

Issue: fixes #665

JakubVanek commented:
Tested on the `4.14.117-ev3dev-2.3.4-ev3` kernel. Current ev3dev-lang-java stops only in the beginning. With this patch, the motor is stopped on each command.

```java
package issues;

import ev3dev.actuators.lego.motors.EV3LargeRegulatedMotor;
import lejos.hardware.port.MotorPort;
import lejos.utility.Delay;

public class Issue667 {
    public static void main(String[] args) {
        EV3LargeRegulatedMotor m = new EV3LargeRegulatedMotor(MotorPort.C);
        m.setSpeed(360);

        m.forward();
        Delay.msDelay(1000);
        m.stop();
        Delay.msDelay(1000);

        m.forward();
        Delay.msDelay(1000);
        m.flt();
        Delay.msDelay(1000);

        m.forward();
        Delay.msDelay(1000);
        m.brake();
        Delay.msDelay(1000);

        m.forward();
        Delay.msDelay(1000);
        m.coast();
        Delay.msDelay(1000);

        m.forward();
        Delay.msDelay(1000);
        m.hold();
        Delay.msDelay(1000);

    }
}
```

jabrena commented:
**This is not an error.**

When you design your robot, you model the stop strategy once.
In the runtime, you don´t change the strategy.

jabrena commented:
Next week, I will remove the dev branch for this PR.

JakubVanek commented:
Hi,

however, according to the quoted documentation:
>  Set the motor into float mode. **This will stop the motor** without braking and the position of the motor will not be maintained.

I agree that `set the mode` and then `perform the actual stop` is a valid model - but this is not how it was implemented on leJOS. In leJOS, both steps are merged together. This is the only issue - there is no blocker to doing it like leJOS did. The current ev3dev-lang-java approach works, but it is not backward-compatible.

An alternative solution would be to rename the existing methods:
* `setStopMode()` - this would represent the current `flt()`, `coast()`, `brake()`, `hold()` etc. behavior,
* `doStop()` or `stop()` - this would be represent the current `stop()` behavior.

From the linguistic point of view, `brake()` or `stop()` or `doStop()` feels like imperative - *hey, robot, start braking now* (which does not happen now). On the other hand, using `setStopMode()` would signal that this is not the actual action - it just sets the mode and does not trigger the stop.

I have an argument for a change. If one is driving a differential wheeled robot in a maze, from time to time it is necessary to align the robot to some solid object in the maze. When the robot hits the object, it is necessary to perform a coast-stop to let the robot construction "decompress" and idle a second after the hit. On the other hand, when stopping in the middle of the maze, it is necessary to use the hold-stop to not lose the position knowledge (no LIDAR or camera is available with stock sensors, dead-reckoning could be the only way). In the end, it is just a difference between the two following sets of functions:

Current leJOS way:
```java
public void afterTouchPressed() {
    motorL.flt();
    motorR.flt();
}
public void stopInMaze() {
    motorL.stop();
    motorR.stop();
}
```
Current ev3dev-lang-java way:
```java
public void afterTouchPressed() {
    motorL.coast(); // or motorL.flt() !
    motorR.coast(); // or motorR.flt() !
    motorL.stop();
    motorR.stop();
}
public void stopInMaze() {
    motorL.brake();
    motorR.brake();
    motorL.stop();
    motorR.stop();
}
```

On the other hand, if the `setStopMode()` idea is realized, I think I would have no problem with writing it like the following code. It is then even clearer what the code does.

```java
public void afterTouchPressed() {
    motorL.setStopMode(StopMode.COAST);
    motorR.setStopMode(StopMode.COAST);
    motorL.doStop();
    motorR.doStop();
}
public void stopInMaze() {
    motorL.setStopMode(StopMode.HOLD);
    motorR.setStopMode(StopMode.HOLD);
    motorL.doStop();
    motorR.doStop();
}
```

Regards,

Jakub

jabrena commented:
Hi Jakub, 

reviewing again the Interpretation of `flt()` you are right.

Take a look this article, the explanation about both behaviour is clear:
http://thetechnicgear.com/2014/05/tutorial-using-motors-draw-square-mindstorms-ev3-lejos/

> flt and stop
> So eventually I am sure you will want to stop the motor and we have two options.
> 
> The flt method will stop the current to the motor and will left it floating.
> 
> motor.flt()
> 
> The stop method on the other hand will actually brake and stop the motor.
> 
> motor.stop()
> 
> Both methods have an optional boolean parameter immediateReturn.

Tomorrow, I will accept the PR.

JakubVanek commented:
Thanks!

Jakub

JakubVanek commented on src/main/java/ev3dev/actuators/lego/motors/BaseRegulatedMotor.java:
```diff
@@ -165,18 +165,18 @@ public void backward(){
-    public void flt(boolean b) {
-        this.flt();
+    public void flt(boolean immediateReturn) {
+        doStop(COAST, immediateReturn);
```
This is consistent with leJOS:
flt() sends a move command to the regulator class: https://sourceforge.net/p/lejos/ev3/code/ci/master/tree/ev3classes/src/lejos/hardware/motor/BaseRegulatedMotor.java#l171
Kernel regulator interface handles it here: https://sourceforge.net/p/lejos/ev3/code/ci/master/tree/ev3classes/src/lejos/internal/ev3/EV3MotorPort.java#L427
```
float speed = 0,
int acceleration = <current acceleration>,
int limit = no limit,
boolean hold = FALSE,
boolean waitComplete = <forwarded>
```

JakubVanek commented on src/main/java/ev3dev/actuators/lego/motors/BaseRegulatedMotor.java:
```diff
@@ -165,18 +165,18 @@ public void backward(){
     @Override
     public void flt() {
-        this.setStringAttribute(STOP_COMMAND, COAST);
+        flt(true);
```
This is also consistent with leJOS:
https://sourceforge.net/p/lejos/ev3/code/ci/master/tree/ev3classes/src/lejos/hardware/motor/BaseRegulatedMotor.java#l161
It even seems that `flt()` was asynchronous by default even previously.

JakubVanek replied:
Nope, there is a compatibility break - the true is negated in leJOS into false. This means that leJOS was waiting for a stop by default.

JakubVanek replied:
This is consistent with leJOS:
https://sourceforge.net/p/lejos/ev3/code/ci/master/tree/ev3classes/src/lejos/hardware/motor/BaseRegulatedMotor.java#l183
```
float speed = 0,
int acceleration = <acceleration>,
int limit = <no limit>,
boolean hold = FALSE,
boolean waitComplete = TRUE
```

JakubVanek replied:
There is the change mentioned in the ADR - waitComplete is false by default now.

JakubVanek commented on src/main/java/ev3dev/actuators/lego/motors/BaseRegulatedMotor.java:
```diff
@@ -165,18 +165,18 @@ public void backward(){
     @Override
     public void coast() {
-        this.setStringAttribute(STOP_COMMAND, COAST);
+        doStop(COAST, true);
```
This is a ev3dev-addded function using the new common implemetation.

JakubVanek commented on src/main/java/ev3dev/actuators/lego/motors/BaseRegulatedMotor.java:
```diff
@@ -186,7 +186,7 @@ public void coast() {
      */
     public void brake() {
-        this.setStringAttribute(STOP_COMMAND, BRAKE);
+        doStop(BRAKE, true);
```
This is a ev3dev-addded function using the new common implemetation.

This will not work on BrickPi - it probably can´t physically short the motor pins, so it can´t stop the motor with the brake stop action.

JakubVanek commented on src/main/java/ev3dev/actuators/lego/motors/BaseRegulatedMotor.java:
```diff
@@ -195,7 +195,7 @@ public void brake() {
     @Override
     public void hold() {
-        this.setStringAttribute(STOP_COMMAND, HOLD);
+        doStop(HOLD, true);
```
This is a ev3dev-addded function using the new common implemetation.

JakubVanek commented on src/main/java/ev3dev/actuators/lego/motors/BaseRegulatedMotor.java:
```diff
@@ -206,16 +206,31 @@ public void hold() {
-        for (RegulatedMotorListener listener : listenerList) {
-            listener.rotationStopped(this, this.getTachoCount(), this.isStalled(), System.currentTimeMillis());
-        }
+        stop(true);
```
This is consistent with leJOS:
https://sourceforge.net/p/lejos/ev3/code/ci/master/tree/ev3classes/src/lejos/hardware/motor/BaseRegulatedMotor.java#l183
```
float speed = 0,
int acceleration = <acceleration>,
int limit = <no limit>,
boolean hold = TRUE,
boolean waitComplete = TRUE
```

JakubVanek replied:
However there is the change mentioned in the ADR - waitComplete is false by default now.

JakubVanek commented on src/main/java/ev3dev/actuators/lego/motors/BaseRegulatedMotor.java:
```diff
@@ -206,16 +206,31 @@ public void hold() {
     @Override
-    public void stop(boolean b) {
+    public void stop(boolean immediateReturn) {
+        doStop(HOLD, immediateReturn);
```
This is consistent with leJOS:
https://sourceforge.net/p/lejos/ev3/code/ci/master/tree/ev3classes/src/lejos/hardware/motor/BaseRegulatedMotor.java#l196
```
float speed = 0,
int acceleration = <acceleration>,
int limit = <no limit>,
boolean hold = TRUE,
boolean waitComplete = <forwarded>
```

JakubVanek commented on src/main/java/ev3dev/actuators/lego/motors/BaseRegulatedMotor.java:
```diff
@@ -206,16 +206,31 @@ public void hold() {
+     * @param mode One of BRAKE, COAST and HOLD string constants.
+     * @param immediateReturn Whether the function should busy-wait until the motor stops reporting the 'running' state.
+     */
+    private void doStop(String mode, boolean immediateReturn) {
```
This is the common implementation.

JakubVanek commented on src/main/java/ev3dev/actuators/lego/motors/BaseRegulatedMotor.java:
```diff
@@ -206,16 +206,31 @@ public void hold() {
+
+        if (!immediateReturn) {
+            waitComplete();
+        }
```
This is a new code to honor the `immediateReturn` parameter. 

This is how it is implemented on leJOS:
1. https://sourceforge.net/p/lejos/ev3/code/ci/master/tree/ev3classes/src/lejos/internal/ev3/EV3MotorPort.java#l427 newMove() -> waitComplete()
2. https://sourceforge.net/p/lejos/ev3/code/ci/master/tree/ev3classes/src/lejos/internal/ev3/EV3MotorPort.java#l379 waitComplete() body

JakubVanek commented on src/test/java/fake_ev3dev/ev3dev/actuators/lego/motors/FakeEV3LargeRegulatedMotor.java:
```diff
@@ -54,6 +54,13 @@ public FakeEV3LargeRegulatedMotor(EV3DevPlatform ev3DevPlatform) throws IOExcept
+                        LEGO_TACHO_PATH + "/" +
+                        MOTOR1 + "/" +
+                        MOTOR_STOP_ACTION);
+        createFile(stop_action);
```
This fixes unit tests.

JakubVanek commented:
I have added comments to the code, linking to the equivalent places in the leJOS souce code.

I do not remember why I have changed the default from synchronous stop & flt & etc. to asynchronous ones. This is also a breakage (however now at least documented). I can change the PR to make it behave the same way as it was on leJOS. I think I did this because ev3dev-lang-java has previously always behaved like `immediateReturn=true` (because the wait was not implemented).

JakubVanek commented:
Experiment:
```java
package example;

import ev3dev.actuators.lego.motors.EV3MediumRegulatedMotor;
import lejos.hardware.port.MotorPort;
import lejos.robotics.RegulatedMotor;
import lejos.utility.Delay;
import lombok.extern.slf4j.Slf4j;

@Slf4j
public class StopTest {
    public static void main(String[] args) {
        LOGGER.info("Initializing...");
        RegulatedMotor motor = new EV3MediumRegulatedMotor(MotorPort.B);
        motor.setSpeed(720);

        LOGGER.info("coast()");
        motor.forward();
        Delay.msDelay(500);
        motor.coast();
        LOGGER.info("-> ok");
        Delay.msDelay(3000);

        LOGGER.info("brake()");
        motor.forward();
        Delay.msDelay(500);
        motor.brake();
        LOGGER.info("-> ok");
        Delay.msDelay(3000);

        LOGGER.info("hold()");
        motor.forward();
        Delay.msDelay(500);
        motor.hold();
        LOGGER.info("-> ok");
        Delay.msDelay(3000);

        LOGGER.info("flt(true)");
        motor.forward();
        Delay.msDelay(500);
        motor.flt(true);
        LOGGER.info("-> ok");
        Delay.msDelay(3000);

        LOGGER.info("flt(false)");
        motor.forward();
        Delay.msDelay(500);
        motor.flt(false);
        LOGGER.info("-> ok");
        Delay.msDelay(3000);

        LOGGER.info("stop(true)");
        motor.forward();
        Delay.msDelay(500);
        motor.stop(true);
        LOGGER.info("-> ok");
        Delay.msDelay(3000);

        LOGGER.info("stop(false)");
        motor.forward();
        Delay.msDelay(500);
        motor.stop(false);
        LOGGER.info("-> ok");
        Delay.msDelay(3000);
    }
}
```

JakubVanek commented:
Experiment verified.
